---
title: "Module2_quiz"
author: "OHfromH2O"
date: '2022-06-25'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### **Q1. What is the GC content of “chr22” in the “hg19” build of the human genome?**
### Tip: The reference genome includes “N” bases; you will need to exclude those.
```{r echo=TRUE, results='hide', message=FALSE, warning=FALSE}
library(AnnotationHub)
library(GenomicRanges)
library(GenomicFeatures)
library(GenomeInfoDb)
library(Biostrings)
library(BSgenome)
library(rtracklayer)
library(Rsamtools)
```
```{r echo=TRUE, message=FALSE, warning=FALSE}
available.genomes()
```
```{r echo=TRUE, eval=FALSE, message=FALSE, warning=FALSE}
# download the genome data
library(BiocManager)
BiocManager::install("BSgenome.Hsapiens.UCSC.hg19")
```
```{r echo=TRUE, message=FALSE, warning=FALSE}
# load the genome data
library("BSgenome.Hsapiens.UCSC.hg19")

# calculate total bases on chr22
alphabet_frequency_with_N <- alphabetFrequency(Hsapiens$chr22)
total_bases <- sum(alphabet_frequency_with_N[c('A','G','C','T')])

# count GC bases on chr22  
GC_bases <- sum(alphabet_frequency_with_N[c('G','C')])

#calculate GC ratio
GC_ratio <- GC_bases/total_bases
GC_ratio 
```


### 2. Background: In the previous assessment we studied H3K27me3 “narrowPeak” regions from the H1 cell line (recall that the Roadmap ID for this cell line is “E003”). We want to examine whether the GC content of the regions influence the signal; in other words wether the reported results appear biased by GC content.  
### **Q2. What is mean GC content of H3K27me3 “narrowPeak” regions from Epigenomics Roadmap from the H1 stem cell line on chr 22?**
### Clarification: Compute the GC content for each peak region as a percentage and then average those percentages to compute a number between 0 and 1.
```{r echo=TRUE, message=FALSE, warning=FALSE}
# query H3K4me3 H1 cell line as narrowpeak from AnnotationHub
ah <- AnnotationHub()
ah_H3K27me3 <- query(ah, c("H3K27me3", "narrowpeak", "E003"))
ah_H3K27me3_record <- ah_H3K27me3[["AH29892"]]

# extract chr 22 and create a views object
H3K27me3_chr22 <- subset(ah_H3K27me3_record, seqnames == "chr22")
H3K27me3_chr22_views <- Views(Hsapiens, H3K27me3_chr22)

# calculate mean of GC contents
GC_contents <- letterFrequency(H3K27me3_chr22_views, "GC", as.prob = TRUE)
GCs_mean <- mean(GC_contents)
GCs_mean
```


### 3. The “narrowPeak” regions includes information on a value they call “signalValue”. 
### **Q3. What is the correlation between GC content and “signalValue” of these regions (on chr22)?**
```{r echo=TRUE, message=FALSE, warning=FALSE}
signal_value <- H3K27me3_chr22$signalValue
cor(signal_value, GC_contents)
```


### 4. The “narrowPeak” regions are presumably reflective of a ChIP signal in these regions. To confirm this, we want to obtain the “fc.signal” data from AnnotationHub package on the same cell line and histone modification. This data represents a vector of fold-change enrichment of ChIP signal over input.
### **Q4. What is the correlation between the “signalValue” of the “narrowPeak” regions and the average “fc.signal” across the same regions?**
### Clarification: First compute the average “fc.signal” for across each region, for example using “Views”; this yields a single number of each region. Next correlate these numbers with the “signalValue” of the “narrowPeaks”.
```{r echo=TRUE, message=FALSE, warning=FALSE}
ah <- AnnotationHub()
ah_H3K27me3_fc <- query(ah, c("H3K27me3", "E003", "fc.signal"))
ah_H3K27me3_record <- ah_H3K27me3_fc[["AH32033"]]

# subset data on chr22, using import function
gr22 <- GRanges(seqnames = "chr22", ranges = IRanges(start = start(Hsapiens$chr22), end = end(Hsapiens$chr22)))
H3K27me3_fc_gr <- import(ah_H3K27me3_record, which = gr22, as = "Rle")
H3K27me3_fc_gr22 <- H3K27me3_fc_gr$chr22

# view fc.signal data
fc.signal <- Views(H3K27me3_fc_gr22, start = start(H3K27me3_chr22), end = end(H3K27me3_chr22))

# calculate the correlation between the average of fc.signal and signalValue
fc.signal_mean <- mean(fc.signal)
cor(fc.signar_mean, signal_value)
```


### **Q5. Referring to the objects made and defined in the previous question. How many bases on chr22 have an fc.signal greater than or equal to 1?**
```{r echo=TRUE, message=FALSE, warning=FALSE}
sum(H3K27me3_fc_gr22 >= 1)
```


### 6. The H1 stem cell line is an embryonic stem cell line, a so-called pluripotent cell. Many epigenetic marks change upon differentiation. We will examine this. We choose the cell type with Roadmap ID “E055” which is foreskin fibroblast primary cells. We will use the “fc.signal” for this cell type for the H3K27me3 mark, on chr22. We now have a signal track for E003 and a signal track for E055. We want to identify regions of the genome which gain H3K27me3 upon differentiation. These are regions which have a higher signal in E055 than in E003. To do this properly, we would need to standardize (normalize) the signal across the two samples; we will ignore this for now.  
### **Q6. Identify the regions of the genome where the signal in E003 is 0.5 or lower and the signal in E055 is 2 or higher.**
### Tip: If you end up with having to intersect two different Views, note that you will need to convert the Views to IRanges or GRanges first with ir<-as(vi,"IRanges").
```{r echo=TRUE, message=FALSE, warning=FALSE}
# query the record
ah_H3K27me3_E055 <- query(ah, c("H3K27me3", "E055"))
ah_H3K27me3_E055
ah_H3K27me3_E055_record <- ah_H3K27me3_E055[["AH32470"]]

# subset by chr22, using import
gr22 <- GRanges(seqnames = "chr22", ranges = IRanges(start = start(Hsapiens$chr22), end = end(Hsapiens$chr22)))
H3K27me3_fc_gr_E055 <- import(ah_H3K27me3_E055_record, which = gr22, as = "Rle")
H3K27me3_fc_gr22_E055 <- H3K27me3_fc_gr_E055$chr22

# identify the region
subset(H3K27me3_fc_gr22,)
region_E003 <- as(slice(H3K27me3_fc_gr22, upper = 0.5), "IRanges")
region_E055 <- as(slice(H3K27me3_fc_gr22_E055, lower = 2), "IRanges")
signal_region <- intersect(region_E003, region_E055)
sum(width(signal_region))
```


### 7. CpG Islands are dense clusters of CpGs. The classic definition of a CpG Island compares the observed to the expected frequencies of CpG dinucleotides as well as the GC content. Specifically, the observed CpG frequency is just the number of “CG” dinucleotides in a region. The expected CpG frequency is defined as the frequency of C multiplied by the frequency of G divided by the length of the region.
### **Q7. What is the average observed-to-expected ratio of CpG dinucleotides for CpG Islands on chromosome 22?**
```{r echo=TRUE, message=FALSE, warning=FALSE}
# query the CpG
ah_hg <- subset(ah, species == "Homo sapiens")
ah_hg_CpG <- query(ah_human, "CpG Islands")
ah_hg_CpG
ah_hg_CpG_record <- ah_hg_CpG[["AH5086"]]

# subset data on chr22 and create a views object
ah_hg_CpG_chr22 <- subset(ah_hg_CpG_record, seqnames == "chr22")
ah_hg_CpG_chr22_views <- Views(Hsapiens, ah_hg_CpG_chr22)

# calculate observed "GC" ratio = GC_freq / total length
observed_GC <- dinucleotideFrequency(ah_hg_CpG_chr22_views)[,7]/width(ah_hg_CpG_chr22_views)

# calculate expected GC ratio = (G_freq + C_freq) / total length
G_freq <- letterFrequency(ah_hg_CpG_chr22_views, "G")
C_freq <- letterFrequency(ah_hg_CpG_chr22_views, "C")
expected_GC <- (G_freq/width(ah_hg_CpG_chr22_views))*(C_freq/width(ah_hg_CpG_chr22_views))

# calculate the average observed-to-expected ratio of CpG dinucleotides
CpG_Islands <- mean(observed_GC/expected_GC)
CpG_Islands
```


### 8. A TATA box is a DNA element of the form “TATAAA”. Around 25% of genes should have a TATA box in their promoter. We will examine this statement. 
### **Q8. How many TATA boxes are there on chr 22 of build hg19 of the human genome?** 
### Clarification: You need to remember to search both forward and reverse strands.
```{r echo=TRUE, message=FALSE, warning=FALSE}
TATA_boxes <- countPattern("TATAAA", Hsapiens$chr22) + countPattern("TATAAA", reverseComplement(Hsapiens$chr22))
TATA_boxes
```


### **Q9. How many promoters of transcripts on chromosome 22 containing a coding sequence, contains a TATA box on the same strand as the transcript?**
### Clarification: Use the TxDb.Hsapiens.UCSC.hg19.knownGene package to define transcripts and coding sequence. Here, we defined a promoter to be 900bp upstream and 100bp downstream of the transcription start site.
```{r echo=TRUE, eval=FALSE, message=FALSE, warning=FALSE}
# download and load the data
library(BiocManager)
BiocManager::install("TxDb.Hsapiens.UCSC.hg19.knownGene")
```
```{r echo=TRUE, message=FALSE, warning=FALSE}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
```
```{r echo=TRUE, message=FALSE, warning=FALSE}
# create Granges corresponding to chr22
gr22 <- GRanges(seqnames = "chr22", ranges = IRanges(start = start(Hsapiens$chr22), end = end(Hsapiens$chr22)))

# find coding sequences on chr 22
cds <-  subsetByOverlaps(cdsBy(txdb, by = "tx"), gr22)

# find transcripts on chr 22
txs <- subsetByOverlaps(transcripts(txdb, columns= c("tx_id", "tx_name")), gr22)

# find overlaps between coding sequences and transcripts
proms_cds <- findOverlaps(cds, txs)

# subset the transcripts of txdb to only include the ones from the coding sequences.
list_id <- names(cds)
test_subset <- subset(txs, tx_id %in% list_id)

# get their promoters and change to view objects
proms <- promoters(test_subset, upstream = 900, downstream = 100)
proms_cds_view <- Views(Hsapiens, proms[unique(queryHits(proms_cds))])

# count overlaps between the pattern (TATA) and promoters
plus_strand <- proms_cds_view[strand(proms_cds_view) == "+"]
minus_strand <- proms_cds_view[strand(proms_cds_view) == "-"]
count <- sum(vcountPattern("TATAAA", DNAStringSet(plus_strand))) + sum(vcountPattern("TTTATA", DNAStringSet(minus_strand)))
count
```


### 10. It is possible for two promoters from different transcripts to overlap, in which case the regulatory features inside the overlap might affect both transcripts. This happens frequently in bacteria. 
### **Q10. How many bases on chr22 are part of more than one promoter of a coding sequence?**
### Clarification: Use the TxDb.Hsapiens.UCSC.hg19.knownGene package to define transcripts and coding sequence. Here, we define a promoter to be 900bp upstream and 100bp downstream of the transcription start site. In this case, ignore strand in the analysis.
```{r echo=TRUE, message=FALSE, warning=FALSE}
# calculate transcript lengths
txs_len_chr22 <- transcriptLengths(txdb, with.cds_len = TRUE)
txs_len_chr22 <- txs_len_chr22[txs_len_chr22$cds_len > 0,]

# find promoters which overlap from different transcript
txs_olap <- proms[mcols(proms)$tx_id %in% txs_len_chr22$tx_id]
result = sum(coverage(txs_olap) > 1)
result["chr22"]
```


```{r}
sessionInfo()
```